### DISCUSSION TITLE
Verifiable Credentials URI Specification Documentation

### DISCUSSION THREAD

**Discussion by ChakshuGautam** [Sunbird-RC/community#227] [Ideas & Enhancements]
I was looking at the existing open source VC to QR Code implementation and found this whitepaper - https://github.com/Path-Check/paper-cred. They included COWIN as a example as well - https://github.com/Path-Check/paper-cred/blob/main/payloads/cowin.1.md 

I found this to be a really exhaustive for a leaner data representation than JSON. I have a fair bit of understanding on how RC is doing this but failed to find a similar doc. Does the Sunbird VC provide a similar doc for the QR code generation? Would love to contribute on the docs front for this.

**Comment by pramodkvarma** [comment]
I don't think RC has taken an opinionated position, but, as I understand we have implemented few options (documentation may be missing). In addition, we can also  align PathCheck as an option for adopters. But adopters may choose other libraries as well.

 @dileepbapat @tejash-jl can you help?

**Reply by dileepbapat** [reply]
By Default VC json is used as payload in QR code after compressing with standard zip algorithm. However other format support can be added.

**Reply by pramodkvarma** [reply]
Thanks @dileepbapat 

@chakshugautam do you want to try implementing pathcheck option? That'd be wonderful!

**Reply by ChakshuGautam** [reply]
Yes yes. Thanks. 

We implemented this in one of our programs. Will try to raise a PR later this week. :)

**Reply by pramodkvarma** [reply]
Awesome! You can have @dileepbapat review the PR. Thanks a lot.

**Reply by ChakshuGautam** [reply]
If I am not wrong, currently the certificates are a snapshot of a point in time data in Sunbird RC and verification is of the hash (the data could have changed later). 

What is Sunbird-RC's opinion of *mutable* vs *immutable* VCs? I really like Dhiway's [implementation - cord](https://github.com/dhiway/cord) which creates an immutable chain of VCs using merkle trees, a history of all changes being part of the API that Cord provides.

@amarts @dileepbapat

**Reply by rahul101001000** [reply]
Dhiway's CORD (based on Proof of Authority, built using Substrate) is indeed a great way to store history of VCs issued.

Maharashtra's Universal Pass uses Print2Block from DocChain.io, a Chennai startup - don't think their software is open source. Sample [Universal Pass](https://block.dmrr.mahaonline.gov.in/report?id=2e63e5ad7e1ae7f4fa8568e8cdb49dec4ac701f9469e7289ba84d1ddb2dc385b)

@ChakshuGautam one suggestion is to have a sample implementation and a detailed blogpost for Sunbird-RC+CORD, similar to the way Amar and Ravi brilliantly documented their Beckn+CORD prototype for agriculture. In that case, blockchain was used to update inventory of produce. Here is a [link](https://becknprotocol.io/enabling-trusted-commerce-using-beckn-protocol-and-blockchain/) to the blog.

**Reply by pramodkvarma** [reply]
Totally agree. CORD team will be more than happy to contribute. If you need connect, let me know. I think they are already part of this project.

**Reply by amarts** [reply]
@ChakshuGautam happy to work together on this. Missed the notification earlier somehow, and hence the delay getting back here. Give me another 3 more days for getting back on the proposal on this, and see how we can proceed in the community on this one.

@smohan-dw tagging you here for the reference.

**Reply by gamemaker1** [reply]
Hi,

I would love to help out with this as well, please do let me know how I can contribute!

I am not as familiar with CORD, but have setup a [Ganache](https://trufflesuite.com/docs/ganache/) blockchain once. Would love it if you could introduce me to CORD as well.

Thanks.

**Reply by pramodkvarma** [reply]
@amarts I can vouch for @gamemaker1 😀 (don't ask his age!), can you please wire him into this effort?

**Reply by gamemaker1** [reply]
Thanks a lot @pramodkvarma, that means a lot. I will do my best to contribute to this :)

**Reply by ChakshuGautam** [reply]
@gamemaker1 hey, discussion happening in the `development` channel here - https://discord.gg/V8AqufZu. Please join.

**Reply by gamemaker1** [reply]
Thanks a lot, I have joined the channel and will read through all the messages today!

**Reply by gamemaker1** [reply]
Hi,

Here is a small update as to what I have done so far.

Thanks a lot to @amarts and @smohan-dw for the excellent introduction to CORD, it really helped me get started and cleared a few of my initial doubts. Thank you to @ChakshuGautam for helping me understand the credential verification process.

I have followed a couple of tutorials related to substrate (such as [this one](https://docs.substrate.io/tutorials/work-with-pallets/) regarding setting up pallates) and have played around with the [demos](https://github.com/dhiway/cord.js/blob/develop/demo/src/demo-vc.ts) in the [`dhiway/cord.js`](https://github.com/dhiway/cord.js) repo related to verifiable credentials, that have helped me understand CORD and the SDK better. [On a separate note, @amarts @smohan-dw if there are any open issues/todos/features in the CORD repos, I would like to help out]

Here is my proposal to integrate CORD into the Sunbird RC verifiable credentials flow:

- extend the [credential issuer module](https://github.com/Sunbird-RC/sunbird-rc-core/blob/870d6b3b650eae806abc6d2ba1b328f0922a9160/services/certificate-api/src/routes/certificate_controller.js) to hash the credential and add it to a stream on the blockchain.
- extend the [verification UI](https://github.com/Sunbird-RC/demo-certificate-issuance/blob/main/verification/src/components/VerifyCertificate/index.js) to verify the certificate by checking for the hash on the chain.

Please let me know if there is anything you would change/improve in the above proposal.

I would love to work on this with all of you.

Thank you,
Vedant

**Reply by smohan-dw** [reply]
The integration proposal looks good to me. Happy to support the integration efforts.

We are moving the CORD demos to the [cord-demo-scripts](https://github.com/dhiway/cord-demo-scripts) repository to make it easy to get the journey started. The goal of this repo is to have scripts that will help developers to get started with various scenarios using the pre-defined schemas and verifiable credentials.

How about adding issues focused on creating demo scripts using the DESP schemas?

**Reply by gamemaker1** [reply]
Hi everyone,

Sorry for the delay in implementing the integration. I have completed the first step in the proposal:

> - extend the [credential issuer module](https://github.com/Sunbird-RC/sunbird-rc-core/blob/870d6b3b650eae806abc6d2ba1b328f0922a9160/services/certificate-api/src/routes/certificate_controller.js) to hash the credential and add it to a stream on the blockchain.

and have created a PR for the same: [#152](https://github.com/Sunbird-RC/sunbird-rc-core/pull/152). Please do review the PR and let me know if I should make any changes.

Thanks a lot @amarts and @smohan-dw for quickly answering so many of my questions on the CORD Discord server.

**Reply by pramodkvarma** [reply]
This is amazing @gamemaker1 👏👏👏! Btw, is there a design wiki page or something you can share? Does this implementation abstract the specific blockchain aspects and can plug into any blockchain? Is there an abstraction layer? If not, I strongly suggest we create that.
